File: ios/EmojiCode/Cipher/Decoder.swift
import Foundation

public enum Decoder {
    public static func decode(_ input: String, cipher: [String: String]) -> String {
        guard !input.isEmpty else { return "" }

        var reverse = [String: String]()
        for (kana, emoji) in cipher { reverse[emoji] = kana }

        var dakutenReverse = [Character: Character]()
        for (d, base) in KanaMaps.dakuten { dakutenReverse[base] = d }
        var handakutenReverse = [Character: Character]()
        for (h, base) in KanaMaps.handakuten { handakutenReverse[base] = h }

        // Swiftの Character は書記素クラスタとして扱われるので、
        // 絵文字も含めて1単位で正しく取れる。
        let segments: [Character] = Array(input)
        var result = ""
        var i = 0
        while i < segments.count {
            let seg = String(segments[i])
            let next: Character? = i + 1 < segments.count ? segments[i + 1] : nil

            if let kana = reverse[seg], kana.count == 1 {
                let kCh = kana.first!
                if let n = next, KanaMaps.dakutenMarks.contains(n),
                   let dakuten = dakutenReverse[kCh] {
                    result.append(dakuten)
                    i += 2
                    continue
                }
                if let n = next, KanaMaps.handakutenMarks.contains(n),
                   let handakuten = handakutenReverse[kCh] {
                    result.append(handakuten)
                    i += 2
                    continue
                }
                result.append(kana)
            } else if let kana = reverse[seg] {
                result.append(kana)
            } else {
                result.append(seg)
            }
            i += 1
        }
        return result
    }

    public static func containsCipherEmoji(_ input: String, cipher: [String: String]) -> Bool {
        for e in cipher.values where input.contains(e) {
            return true
        }
        return false
    }
}

File: android/app/src/main/java/com/emojicode/app/cipher/Decoder.kt
package com.emojicode.app.cipher

/** 絵文字を元の文字に戻す。 */
object Decoder {
    fun decode(input: String, cipher: Map<String, String>): String {
        if (input.isEmpty()) return ""

        val reverse = HashMap<String, String>(cipher.size)
        for ((kana, emoji) in cipher) reverse[emoji] = kana

        val dakutenReverse = HashMap<Char, Char>()
        for ((d, base) in KanaMaps.DAKUTEN) dakutenReverse[base] = d
        val handakutenReverse = HashMap<Char, Char>()
        for ((h, base) in KanaMaps.HANDAKUTEN) handakutenReverse[base] = h

        val segments = KanaUtils.toGraphemes(input)
        val result = StringBuilder()
        var i = 0
        while (i < segments.size) {
            val seg = segments[i]
            val next = if (i + 1 < segments.size) segments[i + 1] else null
            val kana = reverse[seg]

            if (kana != null && kana.length == 1) {
                val kCh = kana[0]
                if (next != null && next.length == 1 && next[0] in KanaMaps.DAKUTEN_MARKS
                    && dakutenReverse[kCh] != null) {
                    result.append(dakutenReverse[kCh])
                    i += 2; continue
                }
                if (next != null && next.length == 1 && next[0] in KanaMaps.HANDAKUTEN_MARKS
                    && handakutenReverse[kCh] != null) {
                    result.append(handakutenReverse[kCh])
                    i += 2; continue
                }
                result.append(kana)
            } else if (kana != null) {
                result.append(kana)
            } else {
                result.append(seg)
            }
            i++
        }
        return result.toString()
    }

    /** 暗号表のいずれかの絵文字が含まれているか */
    fun containsCipherEmoji(input: String, cipher: Map<String, String>): Boolean {
        for (e in cipher.values) if (input.contains(e)) return true
        return false
    }
}
